libmultilabel/metrics.py

```diff
@@ -31,68 +26,75 @@ class MultiLabelMetrics():
+        for metric in self.metric_stats:
+            self.metric_stats[metric] = 0.
+
     def add_values(self, y_true, y_pred):
         """Add batch of y_true and y_pred.
 
         Args:
             y_true (ndarray): an array with ground truth labels (shape: batch_size * number of classes)
-            y_pred (ndarray): an array with predicted labels (shape: batch_size * number of classes)
+            y_pred (ndarray): an array with predicted label values (shape: batch_size * number of classes)
         """
-        self.y_true.append(y_true)
-        self.y_pred.append(y_pred)
-
-    def eval(self):
-        """Evaluate precision, recall, micro-f1, macro-f1, and P@k/R@k listed in the monitor_metrics."""
-        y_true = np.vstack(self.y_true)
-        y_pred = np.vstack(self.y_pred)
-        report_dict = classification_report(y_true, y_pred > self.metric_threshold, output_dict=True, zero_division=0)
-        result = {
-            'Micro-Precision': report_dict['micro avg']['precision'],
-            'Micro-Recall': report_dict['micro avg']['recall'],
-            'Micro-F1': report_dict['micro avg']['f1-score'],
-            'Macro-F1': report_dict['macro avg']['f1-score'],
-            'Another-Macro-F1': another_macro_f1(
-                macro_prec=report_dict['macro avg']['precision'],
-                macro_rec=report_dict['macro avg']['recall'])  # caml's macro-f1
-        }
-        # add metrics like P@k, R@k to the result dict
+        y_pred_pos = y_pred > self.metric_threshold
+
+        n_eval = len(y_true)
+        self.n_eval += n_eval
+        self.multilabel_confusion_matrix += multilabel_confusion_matrix(y_true, y_pred_pos)
+
         scores = precision_recall_at_ks(y_true, y_pred, top_ks=self.top_ks)
-        precision_recall_monitored = [
-            metric for metric in self.monitor_metrics if metric in scores]
-        result.update({metric: scores[metric] for metric in precision_recall_monitored})
-        self.cached_results = result
-
-    def get_y_pred(self):
-        """Convert 3D array (shape: number of batches * batch_size * number of classes
-        to 2D array (shape: number of samples * number of classes).
-        """
-        return np.vstack(self.y_pred)
+        for metric in self.prec_recall_metrics:
+            self.metric_stats[metric] += (scores[metric] * n_eval)
 
-    def get_metric_dict(self, use_cache=False):
-        """Evaluate or get score dictionary from cache.
+    def get_metric_dict(self):
+        """Get evaluation results."""
 
-        Args:
-            threshold (float, optional): threshold to evaluate precision, recall, and f1 score. Defaults to 0.5.
-            use_cache (bool, optional): return cached results or not. Defaults to False.
-        """
-        if not use_cache:
-            self.eval()
-        return self.cached_results
+        result = {}
+        cm = self.multilabel_confusion_matrix
+        cum_cm = cm.sum(axis=0)
```

Comment:
Usually we use full words if possible, but cum is not. Can use sum instead
cm_sum = cm.sum(...)
tp_sum, fpr_sum, ...